src/treemodel/schema.js

```diff
@@ -0,0 +1,417 @@
+	 *
+	 * @param {String} type Paths' type. Possible values are `ALLOW` or `DISALLOW`.
+	 * @param {Array.<String>} checkPath Path to check.
+	 * @param {String} [attribute] If set, only paths registered for given attribute will be returned.
+	 * @returns {Boolean} `true` if item has any registered matching path, `false` otherwise.
+	 * @private
+	 */
+	_hasMatchingPath( type, checkPath, attribute ) {
+		const itemPaths = this._getPaths( type, attribute );
+
+		checkPath = checkPath.slice();
+
+		// We check every path registered (possibly with given attribute) in the item.
+		for ( let itemPath of itemPaths ) {
+			// We have one of paths registered in the item.
+
+			// Now we have to check every item name from the path to check.
+			for ( let checkName of checkPath ) {
+				// Every item name is expanded to all names of items that item is extending.
+				// So, if on item path, there is an item that is extended by item from checked path, it will
+				// also be treated as matching.
+				const chain = this._schema._extensionChains.get( checkName );
+
+				// Since our paths have to match in given order, we always check against first item from item path.
+				// So, if item path is: B D E
+				// And checked path is: A B C D E
+				// It will be matching (A won't match, B will match, C won't match, D and E will match)
+				if ( chain.indexOf( itemPath[ 0 ] ) > -1 ) {
+					// Every time we have a match, we remove it from `itemPath` so we can still check against first item.
+					itemPath.shift();
+				}
+			}
+
+			// If `itemPath` has no items it means that we removed all of them, so we matched all of them.
+			// This means that we found a matching path.
+			if ( itemPath.length === 0 ) {
+				return true;
+			}
+		}
+
+		// No matching path found.
+		return false;
+	}
+
+	/**
+	 * Custom toJSON method to solve child-parent circular dependencies.
+	 *
+	 * @returns {Object} Clone of this object with the parent property replaced with its name.
+	 */
+	toJSON() {
+		const json = clone( this );
+
+		// Due to circular references we need to remove parent reference.
+		json._schema = '[treeModel.Schema]';
+
+		return json;
+	}
+}
+
+/**
+ * Schema is a run-time created and modified description of which entity in Tree Model is allowed to be inside another
+ * entity. It is checked to verify whether given action can be preformed on Tree Model or whether Tree Model
+ * is in correct state.
+ *
+ * Schema consist of {@link core.treeModel.SchemaItem schema items}, each describing different entity. Entity can be
+ * either a Tree Model element or an abstract group for similar elements. Entities are represented by names. Names
+ * of special/abstract entities should be prefixed by `$` sign.
+ *
+ * Each entity in Schema may have a set of allow/disallow rules. Every rule describes in which entities given entity
+ * can or cannot be.
+ *
+ * Entities can extend other entities. This mechanism allows for grouping entities under abstract names. Whenever a rule
```

Comment:
This paragraph made me think that perhaps extensions could be implemented JS-way.  If one item extends another item, make it a child class of the parent item. The theoretical `getAllowed()` method would be implemented like this:

```
getAllowed() {
    const allowedByParent = super.getAllowed();

    return allowedByParent.concat( this._allowed );
}
```

So the logic of extending one item by another item would be implemented totally inside `SchemaItem`. Perhaps this would clarify the implementation.